### ISSUE TITLE
Make Corpus immutable and hashable

### ISSUE THREAD

**Issue by oadams** [oadams/persephone#128]

**Comment by shuttle1987** [comment]
What use-case is there for enabling this?

**Comment by oadams** [comment]
Okay, I guess the hashability isn't a big deal, but perhaps if the classes hash function was based on the state the original data is in (among other things), then looking at the has of the corpus object used in an experimental pipeline would be useful for reproducability. Implementing the `__hash__()` special method isn't strictly required for this though.

I think the immutability is probably useful for preventing bugs creeping in. Since all the preprocessing happens in `__init__()`, we don't want, say, the list of utterance prefixes to be changed afterwards and then not have the corresponding feature extraction run.

**Comment by shuttle1987** [comment]
Python's `__hash__` dunder method is an internal implementation detail to support the various hash-table backed associative mappings in the language. It's not a cryptographic hash that will always yield the same result given the same input:
```python
janis@janis-workstation:~$ python3
Python 3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> hash("abcdef")
3361618094207257066
>>> exit()
janis@janis-workstation:~$ python3
Python 3.5.2 (default, Nov 23 2017, 16:37:01) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> hash("abcdef")
5563937455588773354
>>> exit()
```

If you want to hash the contents of an object this could be implemented via a method that uses a proper cryptographic hash if this would aid the goals of the project.

**Comment by oadams** [comment]
Yes, that's right. The idea I had was to make Corpora immutable and the API should be consistent with this behaviour. (However, I'm open to counter-arguments to the underlying design choice)
